pkg/imgutil/dockerconfigresolver/dockerconfigresolver.go

```diff
@@ -201,7 +201,7 @@ func NewAuthCreds(refHostname string) (AuthCreds, error) {
 			// - ac.ServerAddress:         "https://index.docker.io/v1/".
 			if !isAuthConfigEmpty(ac) {
 				if authConfigHostname == registry.IndexServer {
-					if ac.ServerAddress != registry.IndexServer {
+					if ac.ServerAddress != "" && ac.ServerAddress != registry.IndexServer {
```

Comment:
nit: To preserve the behaviour of printing the debug log on `ac.ServerAddress == ""`, I think the `if` logic should be something like:

```go
if ac.ServerAddress == "" {
	// This can happen with Amazon ECR: https://github.com/containerd/nerdctl/issues/733
	logrus.Debugf("failed to get ac.ServerAddress for authConfigHostname=%q (refHostname=%q)",
		authConfigHostname, refHostname)
} else if authConfigHostname == registry.IndexServer {
	if ac.ServerAddress != registry.IndexServer {
		return nil, fmt.Errorf("expected ac.ServerAddress (%q) to be %q", ac.ServerAddress, registry.IndexServer)
	}
} else {
	acsaHostname := credentials.ConvertToHostname(ac.ServerAddress)
	if acsaHostname != authConfigHostname {
		return nil, fmt.Errorf("expected the hostname part of ac.ServerAddress (%q) to be authConfigHostname=%q, got %q",
			ac.ServerAddress, authConfigHostname, acsaHostname)
	}
}
```

Reply:
@ericpromislow @jandubois ↑ LGTY?

Reply:
@AkihiroSuda Yes, LGTM, and I've tested that it fixes the repro case specified in https://github.com/containerd/nerdctl/issues/1308#issuecomment-1213554489

Reply:
BTW, @ericpromislow is still away, so you won't get any feedback from him for at least another 10 days.

Reply:
@jandubois Thanks for testing and providing the info

@ktock Please add your commit to this PR (or just open a new PR)?